crates/key-value-aws/src/store.rs

```diff
@@ -0,0 +1,517 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::Result;
+use aws_config::{BehaviorVersion, Region, SdkConfig};
+use aws_credential_types::Credentials;
+use aws_sdk_dynamodb::{
+    config::{ProvideCredentials, SharedCredentialsProvider},
+    operation::{batch_get_item::BatchGetItemOutput, batch_write_item::BatchWriteItemOutput},
+    primitives::Blob,
+    types::{
+        AttributeValue, DeleteRequest, Get, KeysAndAttributes, PutRequest, TransactGetItem,
+        TransactWriteItem, Update, WriteRequest,
+    },
+    Client,
+};
+use spin_core::async_trait;
+use spin_factor_key_value::{log_error, Cas, Error, Store, StoreManager, SwapError};
+
+pub struct KeyValueAwsDynamo {
+    table: Arc<String>,
```

Comment:
That is fair -- the reasoning for `client` not being Arc is that client already wraps an Arc:
```
pub struct Client {
    handle: Arc<Handle>,
}
```

It should be low cost to clone the client without another Arc was my initial thought, but definitely open to thoughts!

Reply:
Yeah, it would have made sense to me if KeyValueAwsDynamo was Clone, but I didn't see a Clone implementation.  And although `Client` is clone, I don't think `async_once_cell::Lazy<Client, std:something::Terrifying<...>>` is, which now makes me wonder if this might be a hangover from a previous iteration?

Reply:
You are quite correct -- cloning these fields is only done in `StoreManager::get` and `StoreManager::summary` and the overall object is never cloned. But, I now realize that function itself returns an Arc -- so this may be unnecessary overhead.

In the `Store` implementation for `AwsDynamoStore` though, creating CAS handles through the `new_compare_and_swap` function made it seem ideal to use Arcs/low cost cloned objects for the client and table to allow creating many parallel CAS handles at low cost. That doesn't preclude containing owned data in the KeyValueAwsDynamo and creating an Arc for table in `StoreManager::get` though, would that be preferable?

Reply:
I agree that the considerations for `AwsDynamoStore` are different - it was specifically this `KeyValueAwsDynamo` struct that was puzzling me.  But yeah if the expectation is that these fields will be repeatedly cloned then it makes sense to `Arc` them once here.  And with `client`, as you say, it's already cheap to await (except for the first time) and clone the result.  Thanks for the patient explanation - I'm happy now.

Reply:
It is certainly possible that this is premature optimization and caching of the store is done in a table higher in the orchestration of key-value factors -- open to tweaking this! I appreciate the discussion :)